### PR TITLE
[MIG] base_vat_optional_vies: Migration to 10.0

### PR DESCRIPTION
**Pull request by chienandalu**
Optional validation of VAT via VIES
===================================

This module extends base_vat module features allowing to know if VIES
validation was passed or not.

cc @Tecnativa

Migrated from: https://github.com/OCA/account-financial-tools/pull/427

### CODE REVIEW

File: base_vat_optional_vies/models/res_partner.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    vatnumber = None
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    vies_passed = fields.Boolean(
+        string="VIES validation passed", readonly=True)
+
+    def __init__(self, pool, cr):
```
Comment (pedrobaeza) [inline base_vat_optional_vies/models/res_partner.py]: Don't do this hack

File: base_vat_optional_vies/models/res_partner.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        super(ResPartner, self).__init__(pool, cr)
+        self._constraints = []
+
+    @api.constrains('vat')
+    def check_vat(self):
+        for partner in self:
+            if (not self.env.context.get('avoid_check_vat') and
```
Comment (pedrobaeza) [inline base_vat_optional_vies/models/res_partner.py]: Call super always. In this case, you can call it when VIES bypass is not activated. Put also a mechanism for avoiding this special check when on test mode and not testing specifically this module. Use a similar technique than https://github.com/OCA/bank-payment/blob/c5b093fb5ef1df31a699e49c90db646db115b44a/account_payment_transfer_reconcile_batch/models/payment_order.py#L28

File: base_vat_optional_vies/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        if res:
+            vat = country_code.upper() + vat_number
+            if self.vat != vat:
+                data['vat'] = vat
+        if data:
+            self.with_context(avoid_check_vat=True).write(data)
+        return res
+
+    # Delete old api constraint defined in base_vat addon
+    @api.multi
```
Comment (pedrobaeza) [inline base_vat_optional_vies/models/res_partner.py]: This hack shouldn't be needed

File: base_vat_optional_vies/tests/test_res_partner.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            self.assertEqual(res_country.upper(), vat_country)
+            self.assertEqual(res_number, vat_number)
+
+    def _test_validate_vat(self, cases):
+        for vat, res_vat, res_vies in cases:
+            self.partner.vat = vat
+            self.assertEqual(self.partner.vat.upper(), res_vat)
+            self.assertEqual(self.partner.vies_passed, res_vies)
+
+    # AEA: Can't use this test in Travis, VIES checking returns always False
```
Comment (pedrobaeza) [inline base_vat_optional_vies/tests/test_res_partner.py]: Remove this if not used, but try again to see. Maybe you can mock the call to VIES checking

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: base_vat_optional_vies/models/res_partner.py [hunk 5]
```diff
@@ -99,7 +95,10 @@ def vies_vat_optional_check(self, country_code, vat_number):
             # Validate against VAT Information Exchange System (VIES)
             # see also http://ec.europa.eu/taxation_customs/vies/
             vat = country_code + vat_number
-            res = vatnumber.check_vies(vat)
+            if not self.env.context.get('test_vies'):
```
Comment (pedrobaeza) [inline base_vat_optional_vies/models/res_partner.py]: You are ignoring config['test_enable'] as specified and still not calling super...

**Comment by chienandalu** [comment]
@pedrobaeza Yes, calling the super there required to make use of the base check methods. Wich is better and more simple anyway. Take a look now.

File: base_vat_optional_vies/models/res_partner.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        if data:
+            self.with_context(avoid_check_vat=True).write(data)
+        return res
+
+    @api.constrains('vat')
+    def check_vat(self):
+        for partner in self:
+            if (not self.env.context.get('avoid_check_vat') and
+                    not partner.parent_id):
+                super(ResPartner, partner).check_vat()
```
Comment (pedrobaeza) [inline base_vat_optional_vies/models/res_partner.py]: I don't get why do you need to still overwrite this if you have already touched the other methods.

**Comment by chienandalu** [comment]
@pedrobaeza It avoids checking again vat here: https://github.com/Tecnativa/account-financial-tools/blob/180d9c01a3e0c57dd8b1495c4af5e20d4bb08940/base_vat_optional_vies/models/res_partner.py#L83
It also avoids check on child contacts (wich I'm not sure if it's necessary/correct)

**Review by pedrobaeza — APPROVED** [review]

**Review by luismontalba — APPROVED** [review]
Tested on runbot

**Pull request merged**